flutter/lib/models/model.dart

```diff
@@ -2052,6 +2121,119 @@ class CanvasModel with ChangeNotifier {
     }
   }
 
+  Timer? _edgeScrollTimer;
+  double _lastEdgeScrollX = 0.0, _lastEdgeScrollY = 0.0;
+  bool _bumpMouseIsWorking = true;
+
+  setEdgeScrollTimer() {
+    _edgeScrollTimer = Timer(
+      const Duration(milliseconds: 10),
+      () => edgeScrollMouse(_lastEdgeScrollX, _lastEdgeScrollY));
+  }
+
+  cancelEdgeScrollTimer() {
+    _edgeScrollTimer?.cancel();
+    _edgeScrollTimer = null;
+  }
+
+  edgeScrollMouse(double x, double y) async {
+    if (size.width == 0 || size.height == 0) {
+      return;
+    }
+
+    if (!_horizontal.hasClients || !_vertical.hasClients) {
+      return;
+    }
+
+    if (_suppressEdgeScroll) {
+      return;
+    }
+
+    // Latch current coordinate in case we're autorepeating
+    _lastEdgeScrollX = x;
+    _lastEdgeScrollY = y;
+
+    // Trigger scrolling when the cursor is close to an edge
+    const double edgeThickness = 120;
+
+    // But not right at the very edge, otherwise it could trigger by accident when
+    // hovering around the window border.
+    const double safeZoneThickness = 20;
+
+    var dxOffset = 0.0;
+    var dyOffset = 0.0;
+
+    Rect activeZone = Rect.fromLTWH(0, 0, size.width, size.height)
+      .deflate(safeZoneThickness);
+
+    if (activeZone.contains(Offset(x, y))) {
+      if (x < edgeThickness) {
+        dxOffset = x - edgeThickness;
+      } else if ((x >= size.width - edgeThickness)) {
+        dxOffset = x - (size.width - edgeThickness);
+      }
+
+      if (y < edgeThickness) {
+        dyOffset = y - edgeThickness;
+      } else if ((y >= size.height - edgeThickness)) {
+        dyOffset = y - (size.height - edgeThickness);
+      }
+    }
+
+    var encroachment = Vector2(dxOffset, dyOffset);
+
+    var scrollPixel = Vector2(
+      _horizontal.position.pixels,
+      _vertical.position.pixels);
+
+    final max = Vector2(
+      _horizontal.position.maxScrollExtent,
+      _vertical.position.maxScrollExtent);
+
+    encroachment.clamp(-scrollPixel, max - scrollPixel);
+
+    if (encroachment.length > 0) {
+      var bumpAmount = -encroachment;
+
+      // Round away from 0
+      bumpAmount.x += bumpAmount.x.sign * 0.5;
+      bumpAmount.y += bumpAmount.y.sign * 0.5;
+
+      var bumpMouseSucceeded =
+        _bumpMouseIsWorking &&
+        (await rustDeskWinManager.call(
+          WindowType.Main,
+          kWindowBumpMouse,
+          {"dx": bumpAmount.x.round(), "dy": bumpAmount.y.round()})).result;
+
+      if (!bumpMouseSucceeded) {
+        // If we can't BumpMouse, then we switch to slower scrolling with autorepeat
+
+        // Don't keep hammering BumpMouse if it's not working.
+        _bumpMouseIsWorking = false;
+
+        // Keep scrolling as long as the user is overtop of an edge. Each time the
+        // timer fires sets a new timer.
+        setEdgeScrollTimer();
+
+        // When autorepeating, scale down the extent of the scroll, otherwise it'll be way too fast.
+        encroachment *= 0.2;
```

Comment:
My testing shows that `0.2` is still fast.

`0.05` seems acceptable to me. What do you think?

Reply:
Mm, interesting. On my end, 0.2 felt exactly right. Ultimately, it will depend on how precise the timer is. The code currently requests a 10ms interval. Maybe on my system, it is in practice getting a longer delay?? Weird that it would be different.

I'm pretty sure 0.05 would be glacially slow on my end... Let me do some testing. Maybe an algorithm that automatically tunes based on feedback from the system high-precision timer is needed.